symengine/derivative.cpp

```diff
@@ -627,81 +662,98 @@ class DiffImplementation
-                   add(mul(polygamma(zero, beta_arg0), diff_beta_arg0),
-                       sub(mul(polygamma(zero, beta_arg1), diff_beta_arg1),
-                           mul(polygamma(zero, add(beta_arg0, beta_arg1)),
-                               add(diff_beta_arg0, diff_beta_arg1)))));
+        apply(beta_arg0);
+        RCP<const Basic> diff_beta_arg0 = result_;
+        apply(beta_arg1);
+        RCP<const Basic> diff_beta_arg1 = result_;
+        result_ = mul(self.rcp_from_this(),
+                      add(mul(polygamma(zero, beta_arg0), diff_beta_arg0),
+                          sub(mul(polygamma(zero, beta_arg1), diff_beta_arg1),
+                              mul(polygamma(zero, add(beta_arg0, beta_arg1)),
+                                  add(diff_beta_arg0, diff_beta_arg1)))));
     }
 
-    static RCP<const Basic> diff(const Set &self, const RCP<const Symbol> &x)
+    void bvisit(const Set &self)
     {
         throw SymEngineException("Derivative doesn't exist.");
     }
 
-    static RCP<const Basic> diff(const Boolean &self,
-                                 const RCP<const Symbol> &x)
+    void bvisit(const Boolean &self)
     {
         throw SymEngineException("Derivative doesn't exist.");
     }
 
-    static RCP<const Basic> diff(const GaloisField &self,
-                                 const RCP<const Symbol> &x)
+    void bvisit(const GaloisField &self)
     {
         GaloisFieldDict d;
         if (self.get_var()->__eq__(*x)) {
             d = self.get_poly().gf_diff();
-            return GaloisField::from_dict(self.get_var(), std::move(d));
+            result_ = GaloisField::from_dict(self.get_var(), std::move(d));
         } else {
-            return GaloisField::from_dict(self.get_var(), std::move(d));
+            result_ = GaloisField::from_dict(self.get_var(), std::move(d));
         }
     }
 
-    static RCP<const Basic> diff(const Piecewise &self,
-                                 const RCP<const Symbol> &x)
+    void bvisit(const Piecewise &self)
     {
         PiecewiseVec v = self.get_vec();
         for (auto &p : v) {
-            p.first = p.first->diff(x);
+            apply(p.first);
+            p.first = result_;
         }
-        return piecewise(std::move(v));
+        result_ = piecewise(std::move(v));
     }
-};
 
-#define IMPLEMENT_DIFF(CLASS)                                                  \
-    RCP<const Basic> CLASS::diff(const RCP<const Symbol> &x) const             \
-    {                                                                          \
-        return DiffImplementation::diff(*this, x);                             \
+    void apply(const Basic &b)
+    {
+        apply(b.rcp_from_this());
+    }
+
+    void apply(const RCP<const Basic> &b)
+    {
+        auto it = visited.find(b);
+        if (it == visited.end()) {
+            b->accept(*this);
+            insert(visited, b, result_);
+        } else {
+            result_ = it->second;
```

Comment:
This might be very slow for some expressions. We need to benchmark it.

Reply:
Whatever is easier for you. I just don't want to merge anything into master that slows things down.

Reply:
I reverted the `don't revisit` part of the PR.